learning/tour-of-beam/frontend/lib/repositories/models/group.dart

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import 'package:json_annotation/json_annotation.dart';
+
+import 'node.dart';
+
+part 'group.g.dart';
```

Comment:
May we version control this file?  For example, the official flutter samples shows version controlling the generated json_serialization along with the base class:
https://github.com/flutter/samples/tree/main/jsonexample/lib/json_serializable

Reply:
Essentially it would be great if someone cloning this repository fresh, without any knowledge, could follow the README and convention to run and test.

Reply:
That link shows `json_serializable`'s own examples, and they are version-controlled as an example for the people to see what is generated.

In practice, it is rare useful to version-control generated files because:
- They can accidentally get behind their originals or get edited manually. Then mere fixing the generation may break the project. See https://github.com/apache/beam/issues/21925
- They make diffs larger.

In our particular setting this adds a burden of maintaining the Apache license comments in those files because it gets deleted all the time, and it was painful with Playground generated test mocks until we got rid of those files in the repository. To set if off, we can add an exception for `./gradlew rat`, but that just adds things to maintain.

We aim to eliminate generated files from version control. We are almost done with this on the frontend with the only controlled files are gRPC protobuf artifacts.

Reply:
I would just advocate to consider an audience who doesn't have the history of decisions to be able to follow a step-by-step instruction to execute in their chosen environment, whether that is local machine or Google Cloud project, for example.

Reply:
I just saw that you are accomplishing with this akvelon:pg_23304_beam-symbols  Thank you :)

Reply:
I changed my mind towards version-controlling the generated code after reflecting on your point of a barrier to newcomers.

I see one more reason to version-control such files. We had been expecting the programmers to have Flutter for generation which is OK. But we also start expecting them to have Python for Beam symbols extraction which is not that reasonable. When we start extracting symbols for other languages, we will likely be relying on users having other compilers installed which will be totally wrong.

So I am thinking of such a plan:

1. Version-control all generated files.
2. Exclude all generated files from Apache Rat check (this really is the most painful thing with their version-controlling as we had been adding licenses manually there).
3. When building a Docker container, delete and re-generate all generated files so that their manual modifications have no effect in deployment and we see errors stemming from that early.
4. Add a Gradle task to delete all generated files so that generation is easy to verify and we can test the generated files' consistency before deployment.
5. Delete and re-generate all generated files in `:playground:frontend:precommit`
6. Add a GitHub post-commit workflow to re-generate files and to compare them with the committed versions. Break if a difference is found. This way the committer will receive an email and will soon make it a habit to run the local pre-commit task.

So the only remaining downside is larger diffs.

What do you say?

Reply:
I agree and thank you again for listening.